File: homework2/java/HeartbeatServer.java
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.nio.charset.StandardCharsets;
import java.util.Random;

public class Server {
    public static void main(String[] args) throws Exception{
        DatagramSocket socket = new DatagramSocket(1234);
        socket.setSoTimeout(100);
        Random random = new Random();
        long startTime = System.nanoTime();
        long endTime = startTime;
        while (true) {
            try {
                byte[] message = new byte[1024];
                DatagramPacket packet = new DatagramPacket(message, message.length);
                socket.receive(packet);
                String str = new String(packet.getData(), packet.getOffset(), packet.getLength());
                endTime = Long.parseLong(str.split(" ")[1]);
                System.out.println(str.split(" ")[0]+" "+(endTime-startTime));
            }catch (Exception e){
                if(startTime == endTime){
                    continue;
                }
                if(System.nanoTime() - endTime > 1_000_000_000){ // 1s
                    System.out.println("Heartbeat pause");
                    break;
                }else{
                    System.out.println("Packet lost");
                }
            }
        }
    }
}

File: homework2/python/HeartbeatServer.py
from socket import *
import random
import time

serverSocket = socket(AF_INET,SOCK_DGRAM)
serverSocket.bind(('127.0.0.1',1234))

startTime = float(time.time())
endTime = startTime

while True:
    try:
        serverSocket.settimeout(0.1)
        message,address = serverSocket.recvfrom(1024)
        message = message.decode()
        rtime = float(message.split()[1])
        endTime = rtime
        ping = float(time.time())-rtime
        print('%s:%.8f'%(message.split()[0],ping))
    except Exception as e:
        if endTime == startTime:
            continue
        if time.time() - endTime>=1.0:
            print('Hearbeat pause')
            break
        else:
            print('Packet lost')